README.md

```diff
@@ -1,39 +1,114 @@
 ```
 
+## Authentication process
+
+![alt text](static/auth_process.png "authentication")
+
+All critical authentication information is in the `Sec-Websocket-Protocol` header.
+Variables in the `Sec-Websocket-Protocol` header are prefixed with an id number and separated by commas. The first variable (prefixed by zero) is always the auth service method.
+
+#### #1: Signup
+
+Example headers:
+`Connection: upgrade`
+`Upgrade: websocket`
+`Sec-WebSocket-Version: 13`
+`Accept: sec-websocket-accept`
+`Sec-WebSocket-Protocol: 0signup,1user_one,2placeholder_password,3user_one@gmail,4+00123456,5true,6true`
+Here in the `Sec-Websocket-Protocol` header we have the signup method as the first variable, followed by the user's desired username, password, email, phone, and the last two booleans refer to if the user agreed to the Terms of Service and Terms of Privacy, respectively.
+
+The signup method validates that the username is unique, create the user with the default role 'user', and on success returns the username and a public id generated for the user.
+
+#### #2: Login
+
+Example headers:
+`Connection: upgrade`
+`Upgrade: websocket`
+`Sec-WebSocket-Version: 13`
+`Accept: sec-websocket-accept`
+`Sec-WebSocket-Protocol: 0login,1user_one,2placeholder_password,32,43849823798,5android`
+Here in the `Sec-Websocket-Protocol` header we have the login method as the first variable, followed by the user's registered username, password, service_code (the service the user is logging into), device id, and device OS.
+
+The login method validates that the password hash matches the registered password for the user, and that the requested service matches the user's role, and on success returns the username, the user's public id, and 2 tokens generated for the user, 1 'user' token and 1 'admin' token. These tokens are to match the requested service when calling Authorize (see below) but will only be valid if the user has the necessary role to use the service.
+
+#### #3 + #4: Route with Authorize as middleware
+
+Example headers:
+`Connection: upgrade`
+`Upgrade: websocket`
+`Sec-WebSocket-Version: 13`
+`Accept: sec-websocket-accept`
+`Sec-WebSocket-Protocol: 0authorize,1user_one,2371a695c-b4c1-47b0-b779-638fdc47b4ac,32,43849823798,5android`
+Here in the `Sec-Websocket-Protocol` header we have the authorize method as the first variable, followed by the user's registered username, user or admin token (will be validated according to the chosen service), service_code (the service the user is being authorized into, must match the token), device id, and device OS.
+
+The authorize middleware is used with every other route, it validates that the token, a valid token must have 3 conditions:
+1. the token must be the one generated for this user on Login.
+2. the token must match the service the user is requesting access to.
+3. the user must have the role requirements to use the token.
+
+## Quickstart development
+
+Install [Firecamp]
+
+[Firecamp]: https://firecamp.io
```

Comment:
Firecamp is made for websockets and default value for headers come out of the box if you import the collections!

There are two complications with it though:
1- It does not yet save request bodies, that's why I specified the in the instructions for the user to get the bodies from the "environment" section, that I found is saved in the collection, so I put them there

2- It rounds large numbers (64 bit and upwards), this took me a very long time to debug, and is a major complication when using it with routes that require the user public_id, since the public_id displayed by firecamp is incorrect

Sadly I only found those complications out when I was almost finished, I ended up adding the collections because I thought it was better than not adding them (it provides some level of utility when starting from scratch, and it is really intuitive to establish websocket connections and send/receive messages)

But it is worth noting that it is a small project with only two contributors [(check it out)](https://github.com/firecamp-io/firecamp) so there is probably a lot of room for improvement

Reply:
> 2- It rounds large numbers (64 bit and upwards), this took me a very long time to debug, and is a major complication when using it with routes that require the user public_id, since the public_id displayed by firecamp is incorrect

It's the nature of javascript. The same happens when I was passing log id to FE in order to locate error messages. But I couldn't find the log id in log message. It turns out the i64 problem of JS. I haven't checked out firecamp yet but I guess it's JS

Reply:
Is configs for firecamp code-generated or hand-written?

Reply:
> 1- It does not yet save request bodies, that's why I specified the in the instructions for the user to get the bodies from the "environment" section, that I found is saved in the collection, so I put them there

Can you check this? The docs say it can save the request body.
https://firecamp.io/docs/clients/websocket/preparing-message-collection